borsh-rs/borsh/src/ser/mod.rs

```diff
@@ -269,22 +269,41 @@ impl BorshSerialize for Box<[u8]> {
 macro_rules! impl_arrays {
     ($($len:expr)+) => {
     $(
-      impl<T> BorshSerialize for [T; $len]
-      where T: BorshSerialize
-      {
-        #[inline]
-        fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
-            for el in self.iter() {
-                el.serialize(writer)?;
+        impl<T> BorshSerialize for [T; $len]
+        where T: BorshSerialize
+        {
+            #[inline]
+            fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+                if T::is_u8() && size_of::<T>() == size_of::<u8>() {
+                    // The code below uses unsafe memory representation from `&[T]` to `&[u8]`.
+                    // The size of the memory should match because `size_of::<T>() == size_of::<u8>()`.
+                    //
+                    // `T::is_u8()` is a workaround for not being able to implement `[u8; *]` separately.
+                    let buf = unsafe { std::slice::from_raw_parts(self.as_ptr() as *const u8, self.len()) };
+                    writer.write_all(buf)?;
```

Comment:
Can we use `AsRef<u8>` instead of unsafe magic?

Reply:
The issue is `T` is just `BorshSerialize`, so `[T; $len]` has only `AsRef<T>`. So we can't cast it without unsafe magic.

Reply:
We need specialization or negative trait bounds (to have two implementations: AsRef and !AsRef); until then there is nothing we can do, I guess